pkg/config/configfile.go

```diff
@@ -0,0 +1,369 @@
+	if cf.watchStop != nil {
+		log.Warningf("Run watcher already running for file: %s", cf.file)
+		return
+	}
+
+	cf.watchStop = make(chan struct{})
+
+	go func() {
+		first := true
+
+		var last time.Time
+		var exists bool
+
+		for {
+			// Each iteration, check for the stop trigger, or wait 10 seconds
+			select {
+			case <-cf.watchStop:
+				return
+			case <-time.After(10 * time.Second):
+			}
+
+			// Query stat on the file, on errors other than exists,
+			// we'll need to log the error, and perhaps limit the retries
+			st, err := cf.Stat()
+			if err != nil && !os.IsNotExist(err) {
+				log.Errorf("Storage Stat Error: %s", err)
+				continue
+			}
+
+			// On first iteration, set exists and last modification time (if applicable)
+			// and flip flag
+			if first {
+				exists = !os.IsNotExist(err)
+				if exists {
+					last = st.ModTime
+				}
+				first = false
+				continue
+			}
+
+			// File does not exist in storage, need to check to see if that is different
+			// from last state check
+			if os.IsNotExist(err) {
+				// check to see if the file has gone from exists to !exists
+				if exists {
+					exists = false
+					cf.onFileChange(ChangeTypeDeleted, nil)
+				}
+				continue
+			}
+
+			// check to see if the file has gone from !exists to exists
+			if !exists {
+				data, err := cf.internalRead(true)
+				if err != nil {
+					log.Warningf("Read() Error: %s\n", err)
+					continue
+				}
+				exists = true
+				last = st.ModTime
+				cf.onFileChange(ChangeTypeCreated, data)
+				continue
+			}
+
+			mtime := st.ModTime
+			if mtime != last {
+				last = mtime
+				data, err := cf.internalRead(true)
+				if err != nil {
+					log.Errorf("Read() Error: %s\n", err)
+					continue
+				}
+				cf.onFileChange(ChangeTypeModified, data)
+			}
+		}
+	}()
+}
+
+// stopWatcher closes the stop channel, returning from the runWatcher go routine. Allows us
+// to remove any polling stat checks on files when there are no change handlers.
+func (cf *ConfigFile) stopWatcher() {
+	if cf.watchStop == nil {
+		return
+	}
+
+	close(cf.watchStop)
+	cf.watchStop = nil
+}
+
+// onFileChange is internally called when the core watcher recognizes a change in the ConfigFile. This
+// method dispatches that change to all added watchers
+func (cf *ConfigFile) onFileChange(changeType ChangeType, newData []byte) {
+	// On change, we copy out the handlers to a separate slice for processing for a few reasons:
+	// 1. We don't want to lock while executing the handlers
+	// 2. Handlers may want to operate on the ConfigFile instance, which would result in a deadlock
+	// 3. Allows us to implement priority sorting outside of the lock as well
+	cf.watchLock.Lock()
+	if len(cf.watchers) == 0 {
+		cf.watchLock.Unlock()
+		return
+	}
+
+	toNotify := make([]*pHandler, len(cf.watchers))
+	copy(toNotify, cf.watchers)
+	cf.watchLock.Unlock()
+
+	sort.Sort(byPriority(toNotify))
+
+	for _, handler := range toNotify {
+		handler.handler(changeType, newData)
+	}
```

Comment:
Getting a little crazy, couldn't we have a race here?
1. get halfway through triggering handlers (loop in "thread" 1)
2. get another data update ("thread" 2)
3. start and finish triggering handlers (loop in "thread" 2)
4. finish triggering handlers (loop in "thread" 1)

Reply:
No, the loop calls `onFileChange`, which does not spawn a new goroutine. The handlers are all executed, then it's returned to the loop. 

Think of it like this: 
Thread 1 loops, if it discovers a change, it notifies all the handlers, then continues to loop. Different files can be executing change handlers concurrently, but each "watcher" go routine is local to a single file. 

The way in which we allow concurrent modification is by copying the handler list (in a lock) before executing the handlers. This allows the handler func to possibly remove itself or add new handlers without deadlocking. What we don't allow is any change to propagate from an external thread, which is why the above is not a race.

Reply:
I think I followed that. If you're confident its not a race, I trust you.